### PR TITLE
Move policy to 2018-03-01 sdk, add support for management groups and subscriptions

### PR DESCRIPTION
**Pull request by mentat9**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [X] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by mentat9** [comment]
I haven't submitted PRs to this repo before. Wondering whether there is anything else I need to do to ensure it's on the "active" list: it's been nearly a week with no visible activity. Also, I find it a little weird that one pre-check is still pending.

**Comment by williexu** [comment]
@mentat9 the Ci has finished: https://travis-ci.org/Azure/azure-cli/builds/433663656?utm_source=github_status&utm_medium=notification
Not sure why github is displaying it as pending. Can you address the failed CI checks?

**Comment by williexu** [comment]
Closes: https://github.com/Azure/azure-cli/issues/7421

**Comment by mentat9** [comment]
Fixed the help example (AFAICT), but the failed tests aren't related to any of my changes: no idea how to fix those. What do you normally do in this circumstance?

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py [hunk 1]
```diff
@@ -317,21 +317,27 @@
                 - name: --rules
                   type: string
                   short-summary: Policy rules in JSON format, or a path to a file containing JSON rules.
+                - name: --management-group
+                  type: string
+                  short-summary: Name of the management group the new policy definition can be assigned in.
+                - name: --subscription-id
```
[the remark this answers is not in the document]
Comment (mentat9) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py]: Can't really do that. It's not the subscription of the user, it's the subscription being managed.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py [hunk 2]
```diff
@@ -371,6 +390,12 @@
                 - name: --definitions
                   type: string
                   short-summary: Policy definitions in JSON format, or a path to a file containing JSON rules.
+                - name: --management-group
+                  type: string
+                  short-summary: Name of management group the new policy set definition can be assigned in.
+                - name: --subscription-id
```
[the remark this answers is not in the document]
Comment (mentat9) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py]: Same.

**Comment by williexu** [comment]
@mentat9 I've restarted the CI to bypass an issue earlier today, please check it again, there are failing tests that need to be addressed.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tjprescott) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py]: `--subscription` isn't intended to be the subscription of the user. It is exactly used for referencing subscriptions other than the user's current one.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py [hunk 3]
```diff
@@ -89,41 +92,63 @@ def load_arguments(self, _):
...
+        c.argument('mode', arg_type=get_enum_type(PolicyMode), options_list=('--mode', '-m'), help='Mode of the new policy definition.', min_api='2016-12-01')
+
+    with self.argument_context('policy definition show', resource_type=ResourceType.MGMT_RESOURCE_POLICY) as c:
+        c.argument('policy_definition_name', arg_type=existing_policy_definition_name_type)
+
+    with self.argument_context('policy definition update', resource_type=ResourceType.MGMT_RESOURCE_POLICY) as c:
+        c.argument('policy_definition_name', arg_type=existing_policy_definition_name_type)
+
+    with self.argument_context('policy definition delete', resource_type=ResourceType.MGMT_RESOURCE_POLICY) as c:
+        c.argument('policy_definition_name', arg_type=existing_policy_definition_name_type)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: These are all redundant and should just be under a single context `policy definition`.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py [hunk 4]
```diff
@@ -89,41 +92,63 @@ def load_arguments(self, _):
...
+    with self.argument_context('policy set-definition create', min_api='2017-06-01-preview', resource_type=ResourceType.MGMT_RESOURCE_POLICY) as c:
+        c.argument('name', options_list=('--name', '-n'), help='Name of the new policy set definition.')
+
+    with self.argument_context('policy set-definition show', min_api='2017-06-01-preview', resource_type=ResourceType.MGMT_RESOURCE_POLICY) as c:
+        c.argument('policy_set_definition_name', arg_type=existing_policy_set_definition_name_type)
+
+    with self.argument_context('policy set-definition update', min_api='2017-06-01-preview', resource_type=ResourceType.MGMT_RESOURCE_POLICY) as c:
+        c.argument('policy_set_definition_name', arg_type=existing_policy_set_definition_name_type)
+
+    with self.argument_context('policy set-definition delete', min_api='2017-06-01-preview', resource_type=ResourceType.MGMT_RESOURCE_POLICY) as c:
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: Same here. This code duplication should be eliminated.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 5]
```diff
@@ -393,17 +393,24 @@ def _resolve_policy_id(cmd, policy, policy_set_definition, client):
...
+def _get_custom_or_builtin_policy(cmd, client, name, subscription_id=None, management_group=None, for_policy_set=False):
     from msrest.exceptions import HttpOperationError
     from msrestazure.azure_exceptions import CloudError
     ErrorResponseException = cmd.get_models('ErrorResponseException')
     policy_operations = client.policy_set_definitions if for_policy_set else client.policy_definitions
+
+    if cmd.supported_api_version(min_api='2018-03-01'):
+        enforce_mutually_exclusive(subscription_id, management_group)
+        if subscription_id:
+            client.config.subscription_id = subscription_id
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: This is not supported in CLI. We do not currently allow the user to make API calls from another subscription without running `az account set --subscription`.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mentat9) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py]: Done.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mentat9) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py]: Done.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py (hunk 3 above)
Comment (mentat9) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: Done.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py (hunk 4 above)
Comment (mentat9) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: Done.

**Comment by mentat9** [comment]
@tjprescott, @williexu, this is ready for merge. Please take a look and sign off or comment. Thanks.

**Comment by mentat9** [comment]
History updated per request, but it doesn't appear to have triggered another build. Also, @tjprescott recall, the recording files will need to be munged in order for a couple of the tests to pass, to bypass subscription scrubbing.

**Comment by mentat9** [comment]
@tjprescott, @williexu, this is my last day before 2 weeks of vacation. Is there any way we can get this merged today? We have multiple customers asking for this.

**Comment by mentat9** [comment]
Sure, I get that. I missed the previous check, which is now fixed. But as I've mentioned above, there will also be a couple of test failures. In my meeting with @tjprescott, he said he needed to fix the recording files that require subscription values that are getting scrubbed causing the failures. There will be two test failures due to this problem. Of course, I'll try to find someone for backup while I'm out, but finishing today is way better.

**Comment by mentat9** [comment]
OK, there are 3 tests that failed due to subscription being scrubbed from the recording files. @tjprescott, please fix those files as soon as you can and resubmit the CI. Thanks in advance.

**Comment by tjprescott** [comment]
Closing in favor of #7716.

**Pull request closed without merging**